nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,635 @@
+            while let Some(result) = tasks.join_next().await {
+                total += 1;
+                let CheckResult {
+                    target,
+                    instance_updated,
+                    check_failure,
+                    error,
+                    ..
+                } = result.expect(
+                    "a `JoinError` is returned if a spawned task \
+                    panics, or if the task is aborted. we never abort \
+                    tasks on this `JoinSet`, and nexus is compiled with \
+                    `panic=\"abort\"`, so neither of these cases should \
+                    ever occur",
+                );
+                let mut metrics = self.metrics.lock().unwrap();
+                let metric = metrics.instance(target);
+                if let Some(up) = instance_updated {
+                    if up.instance_updated {
+                        instances_updated += 1;
+                    }
+
+                    if up.vmm_updated {
+                        vmms_updated += 1;
+                    }
+
+                    if !(up.vmm_updated || up.instance_updated) {
+                        no_change += 1;
+                    }
+                    metric.success(up);
+                }
+                if let Some(reason) = check_failure {
+                    match reason {
+                        CheckFailure::NoSuchInstance => not_found += 1,
+                        CheckFailure::SledAgentUnreachable => {
+                            unreachable_instances += 1
+                        }
+                        _ => sled_agent_errors += 1,
+                    }
+
+                    metric.check_failure(reason);
+                }
+                if let Some(reason) = error {
+                    metric.check_error(reason);
+                    check_errors += 1;
+                }
+            }
+
+            slog::info!(opctx.log, "all instance checks complete";
+                "total_instances" => total,
+                "instances_updated" => instances_updated,
+                "unreachable_instances" => unreachable_instances,
+                "vmms_updated" => vmms_updated,
+                "no_change" => no_change,
+                "not_found" => not_found,
+                "sled_agent_errors" => sled_agent_errors,
+                "check_errors" => check_errors,
+                "pruned_instances" => pruned,
+            );
+            serde_json::json!({
+                "total_instances": total,
+                "instances_updated": instances_updated,
+                "vmms_updated": vmms_updated,
+                "no_change": no_change,
+                "not_found": not_found,
+                "unreachable_instances": unreachable_instances,
+                "sled_agent_errors": sled_agent_errors,
+                "check_errors": check_errors,
+                "pruned_instances": pruned,
+            })
+        }
+        .boxed()
+    }
+}
```

Comment:
I know this was a lot of work, but I'm really grateful you structured things this way -- nice to see the built-in concurrency, and great to see we're plumbing out all this data. Thanks for making this task so observable!

Reply:
Thanks for saying so!